Oscars pagination

EmileSonneveld opened the pull request:
Bigger page size, check multiple pages, dedup for when running in auto tests. Only check next page when needed

bossie commented on openeogeotrellis/opensearch.py:
```diff
@@ -22,11 +22,29 @@ def __init__(self, endpoint: str):
+                resp = requests.get(url=url)
+                resp.raise_for_status()
+                json = resp.json()
+                features = json["features"]
```
Rename variable to `collections`.

bossie commented on openeogeotrellis/opensearch.py:
```diff
@@ -22,11 +22,29 @@ def __init__(self, endpoint: str):
+                json = resp.json()
+                features = json["features"]
+                cache_length_before = len(cache)
+                for f in features:
```
Rename variable to `collection`.

EmileSonneveld replied:
Renamed `features` to `collections`

bossie replied:
I think you renamed the wrong one; so `features` -> `collections` and `f` -> `collection`.

EmileSonneveld replied:
fixed

bossie commented on openeogeotrellis/opensearch.py:
```diff
@@ -22,11 +22,29 @@ def __init__(self, endpoint: str):
+                cache_length_before = len(cache)
+                for f in features:
+                    cache[f["id"]] = f
+                if cache_length_before == len(cache):
```
Clarify that this means: if no new items were added to the cache...

Or at least reverse the comparison because this is confusing.

bossie commented on openeogeotrellis/opensearch.py:
```diff
@@ -22,11 +22,29 @@ def __init__(self, endpoint: str):
+                    # Or could be because we reached the end of the list.
+                    break
+                if "totalResults" in json and json["totalResults"] <= len(cache):
+                    # This shortcut is not always available in test environment.
```
So just remove this check then. Things like `totalResults` have proven to be unreliable in the past so just depend on the actual features returned.

EmileSonneveld replied:
Removed. What a petty. But it should not be too bad on performance.

bossie reviewed (changes requested):
A couple of remarks inline but my biggest issue is that if you're doing paging, you want to have the results returned in a particular order like here: https://github.com/Open-EO/openeo-opensearch-client/blob/60d668f9fb6c5e191831b345fb72736e0dd87146/src/main/scala/org/openeo/opensearch/backends/OscarsClient.scala#L73

I'm not sure what the particular `sortKeys` value you want in this case but none of "title", "id" and"identifier" seem to work; "start" does but I'm not sure what it means exactly.

EmileSonneveld commented:
With `start` it seems to be sorted on properties>date. Not on the published date, is that stable enough?
![image](https://github.com/Open-EO/openeo-geopyspark-driver/assets/2298426/38c4951d-a7e9-4a6b-be12-21aa344a11a9)

https://services.terrascope.be/catalogue//collections?startIndex=1&count=200&sortKeys=start
```javascript
data = JSON.parse(document.querySelector("#jsonFormatterRaw > pre").innerText)
console.table(data["features"].map(x=>x["properties"]))
```

EmileSonneveld commented:
I added the sort key anyway. There are multiple products with the same date tough. I guess that after that, it sorted based on when the product was added to the database.

bossie commented:
> I added the sort key anyway. There are multiple products with the same date tough. I guess that after that, it sorted based on when the product was added to the database.

It's probably me but I don't see how this list is sorted on `date` now.

You could go check with Stijn C. to see if there is a particular `sortKeys` that allows reliable paging or just take the pragmatic approach: remove the paging altogether and increase the page size to e.g. 500.

EmileSonneveld commented:
Stijn C says that by default, features are sorted by ID: https://git.vito.be/projects/BIGGEO/repos/oscars/browse/src/main/java/be/vito/opensearch/elasticsearch/SearchOperation.java#85
To be sure I added a warning when it misbehaves. (Still running tests)
If in any way, a feature got returned multiple times, the one on the last page would be preserved.
I added a log for this, and it did not happen on CI at least
